### ISSUE TITLE
[processor/spanmetrics] Possibility to acces index out of range and crash

### ISSUE THREAD

**Issue by Tenaria** [open-telemetry/opentelemetry-collector-contrib#7250]
**Describe the bug**
It is possible for the span metrics processor to access an index out of range hence causing the collector to crash.

The bug occurs on [this line](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.42.0/processor/spanmetricsprocessor/processor.go#L385) and can happen as the `latencyInMilliseconds` value is searched for in the slice p.latencyBounds. However, the last value in the bounds representing the "infinity bucket" is defined by [`math.MaxInt64`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.42.0/processor/spanmetricsprocessor/processor.go#L49) but the `latencyInMilliseconds` value is a float64 in which the maximum value can exceed the maximum value defined by an int64. This means that the binary search won't find the value and will instead [default to returning the index where the value should be inserted](https://pkg.go.dev/sort#SearchFloat64s) i.e an index that currently does not exist. The index is then accessed on [this line](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.42.0/processor/spanmetricsprocessor/processor.go#L439) at which point an out of bounds error is thrown.

**Steps to reproduce**
Set the end time of a span to the current time and the start time of a span to 0 and the application will crash. (This usually should not happen but is a way to trigger the bug)

**What did you expect to see?**
The last value in the slice should be incremented (the value representing "infinity" in histograms).

**What did you see instead?**
The span metrics processor caused the collector to crash due to accessing an index that does not exist. Error: `panic: runtime error: index out of range [17] with length 17`

**What version did you use?**
Version: v0.42.0

**Comment by jpkrohling** [comment]
cc @albertteoh

**Comment by tbthanh90** [comment]
I am facing the same issue too, do we have plan to fix it soon ? Thanks

**Comment by crobertson-conga** [comment]
I replicate this when EndTime is before StartTime on the spans.  A negative value for duration causes the binary search to overflow and try to find a value greater than the max duration already set.

I think a sane option here might just be to check for the case where the index is out of bounds and drop the data point.  i.e. on line 387 of the current processor do something like
```
	// Binary search to find the latencyInMilliseconds bucket index.
	index := sort.SearchFloat64s(p.latencyBounds, latencyInMilliseconds)
	// Negative durations wrap and so will result in out of bounds panic exception
	if index >= len(p.latencyBounds) {
		index = 0
		latencyInMilliseconds = 1
	}
```

**Comment by crobertson-conga** [comment]
I think this is now fixed with #9891
